src/rs/_internal/canonical_fsm/canonical_fsm.rs

```diff
@@ -160,65 +150,77 @@ impl<TPuzzle: SemiGroupActionPuzzle> CanonicalFSM<TPuzzle> {
         > = IndexedVec::default();
 
         let mut mask_to_state = MaskToState::default();
-        mask_to_state.insert(MoveClassMask(0), CANONICAL_FSM_START_STATE);
-        let mut state_to_mask = StateToMask::new(MoveClassMask(0));
+        mask_to_state.insert(
+            MoveClassMask(vec![false; num_move_classes]),
+            CANONICAL_FSM_START_STATE,
+        );
         // state_to_mask, indexed by state ordinal,  holds the set of move classes in the
         // move sequence so far for which there has not been a subsequent move that does not
         // commute with that move.
-        let mut disallowed_move_classes = StateToMask::new(MoveClassMask(0));
+        let mut state_to_mask = StateToMask::new(MoveClassMask(vec![false; num_move_classes]));
 
         let mut queue_index: CanonicalFSMState = CANONICAL_FSM_START_STATE;
-        while Into::<usize>::into(queue_index) < state_to_mask.0.len() {
+        while usize::from(queue_index) < state_to_mask.0.len() {
             let mut next_state: IndexedVec<MoveClassIndex, CanonicalFSMState> =
                 IndexedVec::new(vec![ILLEGAL_FSM_STATE; num_move_classes]);
 
-            let dequeue_move_class_mask: MoveClassMask = state_to_mask.get(queue_index);
-            disallowed_move_classes.push(MoveClassMask(0));
+            let dequeue_move_class_mask = state_to_mask.get(queue_index).clone();
 
             queue_index += CanonicalFSMState(1);
-            let from_state = queue_index;
 
             for move_class_index in generators.by_move_class.index_iter() {
                 let mut skip = false;
                 // If there's a greater move (multiple) in the state that
                 // commutes with this move's `move_class`, we can't move
                 // `move_class`.
-                skip |= (*dequeue_move_class_mask & *commutes[*move_class_index])
-                    >> (*move_class_index + 1)
-                    != 0;
-                skip |= (*dequeue_move_class_mask & *forbidden_transitions[*move_class_index]) != 0;
-                skip |= ((*dequeue_move_class_mask >> *move_class_index) & 1) != 0;
+                skip |= dequeue_move_class_mask
+                    .0
+                    .iter()
+                    .zip(commutes[*move_class_index].0.iter())
+                    .skip(*move_class_index + 1)
+                    .any(|(&dequeue_move_class, &commute)| dequeue_move_class && commute);
```

Comment:
This is a bit hard to read at a glance, do you think we could do it with a method on `MoveClassMask`?